batchflow/models/tf/layers/conv_block.py

```diff
@@ -159,33 +143,296 @@ def conv_block(inputs, layout='', filters=0, kernel_size=3, name=None,
+                                depth_multiplier=self.depth_multiplier)
+                    if self.filters is None or self.filters == 0:
+                        raise ValueError('filters cannot be None or 0 if layout includes convolutional layers')
+
+                elif layer == 'n':
+                    axis = -1 if self.data_format == 'channels_last' else 1
+                    args = dict(fused=True, axis=axis)
+                    call_args.update({'training': training})
+
+                elif self.C_GROUPS[layer] == 'p':
+                    pool_op = 'mean' if layer == 'v' else self.kwargs.pop('pool_op', 'max')
+                    args = dict(op=pool_op, pool_size=self.pool_size, strides=self.pool_strides, padding=self.padding,
+                                data_format=self.data_format)
+
+                elif self.C_GROUPS[layer] == 'P':
+                    pool_op = 'mean' if layer == 'V' else self.kwargs.pop('pool_op', 'max')
+                    args = dict(op=pool_op, data_format=self.data_format,
+                                keepdims=self.kwargs.get('keep_dims', False))
+
+                elif layer in ['d', 'D']:
+                    if self.dropout_rate:
+                        args = dict(rate=self.dropout_rate)
+                        call_args.update({'training': training})
+                    else:
+                        logger.warning('conv_block: dropout_rate is zero or undefined, so dropout layer is skipped')
+                        skip_layer = True
+
+                elif layer == 'O':
+                    if not self.dropout_rate:
+                        dropout_rate = layer_args.get('dropout_rate')
+                    if not self.kwargs.get('block_size'):
+                        block_size = layer_args.get('block_size')
+                    if dropout_rate and block_size:
+                        args = dict(dropout_rate=dropout_rate, block_size=block_size,
+                                    seed=self.kwargs.get('seed'), data_format=self.data_format,
+                                    global_step=self.kwargs.get('global_step'))
+                        call_args.update({'training': training})
+                    else:
+                        logger.warning(('conv_block/dropblock: dropout_rate or block_size is'
+                                        ' zero or undefined, so dropblock layer is skipped'))
+                        skip_layer = True
+
+                elif layer == 'm':
+                    args = dict(depth=self.kwargs.get('depth'), data_format=self.data_format)
+
+                elif layer in ['b', 'B', 'N', 'X']:
+                    args = dict(factor=self.kwargs.get('factor'),
+                                shape=self.kwargs.get('shape'),
+                                data_format=self.data_format)
+                    if self.kwargs.get('upsampling_layout'):
+                        args['layout'] = self.kwargs.get('upsampling_layout')
+
+                else:
```

Comment:
elif layer not in self.C_LAYERS.keys():
raise ValueError('Unknown layer symbol - %s' % layer)

Reply:
Changed